### PR TITLE
add PushGradient rpc in Go

### PR DESCRIPTION
**Pull request by QiJune**
support async mode first.

### CODE REVIEW

File: elasticdl/pkg/common/embedding_table.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ func (table *EmbeddingTable) GetEmbeddingVector(index int64) *Tensor {
 	}
 	// TODO(qijun) only support zero initializer now
 	newVector := NewVector(table.Dim)
+	newVector.Name = table.Name
```
Comment (wangkuiyi) [inline elasticdl/pkg/common/embedding_table.go]: I see there are multiple assignments to the Name property of vector. How about

```go
newVector := NewVector(table.Dim, table.Name)
```

File: elasticdl/pkg/common/embedding_table.go [hunk 2]
```diff
@@ -35,6 +36,7 @@ func (table *EmbeddingTable) GetEmbeddingVector(index int64) *Tensor {
 func (table *EmbeddingTable) GetEmbeddingVectors(indices []int64) *Tensor {
 	d := []int64{int64(len(indices)), table.Dim}
 	t := NewTensor(d)
+	t.Name = table.Name
```
Comment (wangkuiyi) [inline elasticdl/pkg/common/embedding_table.go]:
```go
t := NewTensor(d, table.Name)
```

File: elasticdl/pkg/main/main.go [hunk 3]
```diff
@@ -32,7 +32,7 @@ func main() {
 	flag.Parse()
 	address := fmt.Sprintf("%s:%d", os.Getenv("MY_POD_IP"), *port)
 	serverDone := make(chan bool)
-	ps.CreateServer(address, serverDone)
+	ps.CreateServer(address, int32(*psID), "SGD", 0.1, serverDone)
```
Comment (wangkuiyi) [inline elasticdl/pkg/main/main.go]: How about 

```go
ps.NewServer("SGD", 0.1).Run(address, psID, serverDone)
```

File: elasticdl/pkg/ps/parameter.go [hunk 4]
```diff
@@ -10,13 +10,17 @@ import (
...
+	InitStatus        bool
 }
 
 // NewParameter creates a parameter instance
 func NewParameter() *Parameter {
 	var p Parameter
 	p.NonEmbeddingParam = make(map[string]*common.Tensor)
 	p.EmbeddingParam = make(map[string]*common.EmbeddingTable)
+	p.Version = 0
+	p.InitStatus = false
```
Comment (wangkuiyi) [inline elasticdl/pkg/ps/parameter.go]: How about let `NewParameter` to call make and assign to `NonEmbeddingParam` and `EmbeddingParam`.  The user could simply call

```go
p := NewParameter(0, false)
```

File: elasticdl/pkg/ps/server.go [hunk 5]
```diff
@@ -2,52 +2,119 @@ package ps
...
+func (s *Server) PushGradient(ctx context.Context, in *pb.PushGradientRequest) (*pb.PushGradientResponse, error) {
+	// TODO(qijun) only support async now
+	var res pb.PushGradientResponse
+	var grads []*common.Tensor
+	for _, gradPB := range in.Gradients {
+		grad := common.DeserializeTensorPB(gradPB)
+		grads = append(grads, grad)
+	}
+	err := s.Opt.ApplyGradients(grads, s.Param)
+	s.Param.Version += int32(1)
```
Comment (mhaoli) [inline elasticdl/pkg/ps/server.go:99]: I find that you specify the specific type for constant in many places, e.g. `int32(1)`, `int64(1)`. Is this required by Go?

File: elasticdl/pkg/ps/server.go [hunk 6]
```diff
@@ -2,52 +2,119 @@ package ps
...
 }
 
-func (s *psServer) PullVariable(ctx context.Context, in *pb.PullVariableRequest) (*pb.PullVariableResponse, error) {
-	// TODO: implement the service.
-	return &pb.PullVariableResponse{}, nil
+// NewServer creates a Server instance
+func NewServer(PSID int32, opt string, lr float32) *Server {
+	var ps Server
+	ps.Param = NewParameter()
+	if opt == "SGD" {
```
Comment (skydoorkai) [inline elasticdl/pkg/ps/server.go]: Better to use a function to create an optimizer according to `opt` and other arguments.

File: elasticdl/pkg/ps/server.go [hunk 7]
```diff
@@ -2,52 +2,119 @@ package ps
...
+		ps.Opt = NewSGDOptimizer(lr)
+	}
+	ps.PSID = PSID
+	ps.lock = &sync.Mutex{}
+	return &ps
+}
+
+// PullVariable pulls variable from server
+func (s *Server) PullVariable(ctx context.Context, in *pb.PullVariableRequest) (*pb.PullVariableResponse, error) {
```
Comment (skydoorkai) [inline elasticdl/pkg/ps/server.go:34]: sync-SGD needs a lock. Probably add a TODO here.

File: elasticdl/pkg/common/embedding_table.go (hunk 1 above)
Comment (QiJune) [reply elasticdl/pkg/common/embedding_table.go]: Done

File: elasticdl/pkg/common/embedding_table.go (hunk 2 above)
Comment (QiJune) [reply elasticdl/pkg/common/embedding_table.go]: Done

File: elasticdl/pkg/ps/server.go (hunk 5 above)
Comment (QiJune) [reply elasticdl/pkg/ps/server.go:99]: Yes, the type of `1` is `int`. And in Go, `int`/`int32`/`int64` are different types. We have to cast it.

File: elasticdl/pkg/ps/server.go (hunk 6 above)
Comment (QiJune) [reply elasticdl/pkg/ps/server.go]: I have set up an [issue](https://github.com/sql-machine-learning/elasticdl/issues/1703). @LiMinghao1994 is working on this.

File: elasticdl/pkg/ps/server.go (hunk 7 above)
Comment (QiJune) [reply elasticdl/pkg/ps/server.go:34]: Done.

File: elasticdl/pkg/ps/parameter.go (hunk 4 above)
Comment (QiJune) [reply elasticdl/pkg/ps/parameter.go]: It seems that the default value of int is 0, bool is false in Go.
Remove these two lines code.

File: elasticdl/pkg/main/main.go (hunk 3 above)
Comment (QiJune) [reply elasticdl/pkg/main/main.go]: Done, it's more clear

**Review by mhaoli — APPROVED** [review]

**Pull request merged**